Mocktail/Mocktail.m

```diff
@@ -97,38 +100,172 @@ - (NSSet *)mockResponses;
     return mockResponses;
 }
 
+/**
+ * Find the set of responses that correspond to a tag (folder/directory label)
+ */
+- (NSSet *)responsesForTag:(NSString*)tag
+{
+    NSSet *mockResponses;
+    @synchronized (_mutableMockResponsesInFolders) {
+        if (_mutableMockResponsesInFolders[tag])
+        {
+            mockResponses = [_mutableMockResponsesInFolders[tag] copy];
+        }
+    }
+    return mockResponses;
+}
+
+/**
+ * Find the tags (folder/directory label) that are relevant to a parsed query string
+ */
+- (NSArray *)tagsMatchingQuery:(NSDictionary*)queries
+{
+    NSMutableArray* matchedTags = [NSMutableArray array];
+    @synchronized (_mutableMockResponsesInFolders) {
+        for(NSString* tag in _mutableMockResponsesInFolders.keyEnumerator)
+        {
+            NSDictionary* match = [self matchQuery:queries withTag:tag];
+            if (match)
+            {
+                [matchedTags addObject:match];
+            }
+        }
+    }
+    return [matchedTags sortedArrayUsingComparator:^NSComparisonResult(id a, id b) {
+        NSDate *first = a[@"num_matches"];
+        NSDate *second = b[@"num_matches"];
+        return [second compare:first];
+    }];;
+}
+
+/**
+ * Calculate number of matches for a given tag (folder/directory label). This will yield {num_matches:<number>,tag:<tag>}, where 'num_matches' acts as the goodness value
+ */
+- (NSDictionary *)matchQuery:(NSDictionary*)queries withTag:(NSString*)tag
```

Comment:
Instead of passing a dictionary back, can we add a completion handler that takes the tuple you were returning in the dictionary?
```
(NSInteger numMatches, NSString *tag)
```
unstructured return values kinda scare me.